### DISCUSSION TITLE
Deprecated `skipStateCheck` flag

### DISCUSSION THREAD

**Discussion by ibrahimqasim-abtrace** [panva/openid-client#706] [Q&A] [answered]
To preface, perhaps this should also be in the `oauth4webapi` discussions.

Is it really necessary for the `skipStateCheck` flag to be deprecated to mark it as a warning? I feel like there are lots of reasons for needing to skip the state check, e.g. any stateless application, and the suggested example of `draft-bradley-oauth-jwt-encoded-state-09` or something similar would make perfect sense, essentially using the passed `state` to match against something and then checking the nonce.

For example, I have a fully serverless application, so how is it possible to maintain state between user-agent requests in order to be able to retrieve the correct `state` parameter to pass to the `expectedState` in the `authorizationCodeGrant`?

I suspect the answer will be to use HTTP cookies or just deal with the deprecated warning (e.g. [here](https://github.com/panva/openid-client/discussions/373#discussioncomment-825382)), but just wanted to check.

**Comment by panva** [comment]
> For example, I have a fully serverless application, so how is it possible to maintain state between requests in order to be able to retrieve the correct state parameter to pass to the expectedState in the authorizationCodeGrant

For example - encrypted cookies, persistant kv storage, anything we generally tend to use in web apps to maintain state. 

If you use state for, well, state and have no other means of csrf or code injection protection (pkce, oidc nonce) and the deprecation warning was the reason you stopped to think about why you shouldn't have the need to use it? Yeah, it's doing the job I envisioned for it.

if you do use cookie based pkce (as you should anyway) then ignore the warning, just make sure your state is signed and cannot be forged.

**Reply by ibrahimqasim-abtrace** [reply]
The `code_verifier` and `nonce` need to stored and looked up so that they can be verified during `authorizationCodeGrant`. How can this be done WITHOUT matching against some user-agent?

With `state`, it could be used as a key to look up the `nonce` value, etc.  I've been using this library for years where the `state` parameter has been used to look up values server-side, and this scenario seems common enough for `skipStateCheck` to not need to be deprecated to flag it as a warning...

Btw I just noticed [this](https://github.com/panva/openid-client/blob/feae0dbe4f99feec0381a166281b57e30d05a1e1/examples/oidc.ts#L40) mentions "state" instead of "nonce".

**Reply by panva** [reply]
> Btw I just noticed [this](https://github.com/panva/openid-client/blob/feae0dbe4f99feec0381a166281b57e30d05a1e1/examples/oidc.ts#L40) mentions "state" instead of "nonce".

Thanks. I'll fix that. 

> How can this be done WITHOUT matching against some user-agent?

It can't and that's the point. But it is the same with state. You must pinpoint any of these to the user agent state somehow.

> where the state parameter has been used to look up values server-side

Unless that state value is bound to the user agent (ie what a client session/cookie do for you) then you're open to attacks. 

There are a lot of reasons why we gravitate towards forcing the use of pkce nowadays, the fact that state is often used to implement vulnerable patterns being one of them.

**Reply by ibrahimqasim-abtrace** [reply]
Yeah I have PKCE set up. But no way to match up the `code_verifier`...
I guess I'll need to set up some HTTP only cookies.

Can you spot any obvious issues with in encrypting the state in a HTTP only cookie in the AS redirect response? That cookie would then automatically be passed in the callback request, and can be matched to the `code_verifier` / `nonce`.

**Reply by panva** [reply]
Well, why not encrypt just the code_verifier in a cookie and forego the use of nonce/state altogether...

**Reply by ibrahimqasim-abtrace** [reply]
The authorization server doesn't support PKCE.

**Reply by panva** [reply]
I see. In that case, you've seen the examples, I recommend you use PKCE anyway because your AS might add support of it at some point. And maybe they do support it but don't indicate it, its use is backwards compatible. And then depending on whether you use oidc use nonce in addition, state otherwise.

**Reply by ibrahimqasim-abtrace** [reply]
Yup, that's exactly the approach I'm taking. However, the question is related to matching up the correct `nonce` / `code_verifier` / `state` on the callback request. I suppose the only way to do this is with some cookie.

**Reply by panva** [reply]
Yeah, an encrypted serialized session cookie or a regular backend stored session with an identifier in a cookie, regardless, an httpOnly; cookie that is cleared when you consume the callback.

**Reply by panva** [reply]
So, to summarize, do you think having the deprecation triggered you to look into the subject and improve your deployment's security posture?

**Reply by ibrahimqasim-abtrace** [reply]
Ha! I suppose it did.

But still, I think having the documentation be there for `skipStateCheck` would have been enough. I can't use cookies yet due to our setup, so will have to use some encrypted state anyway. The main reason it bugged me is it doesn't play nice with linting rules 😅

**Reply by panva** [reply, marked as answer]
> I think having the documentation be there for skipStateCheck would have been enough.

It wouldn't. The entire premise of this discussion was that you assumed to have a valid reason to skip checking the state and the deprecation was getting in the way of using it, where in reality using smth along the lines of `draft-bradley-oauth-jwt-encoded-state-09` is the only acceptable way skipping it and that's only because you'll have to validate it yourself (and frankly nobody implements that anyway). The deprecation did exactly as I hoped it would in this particular case.

FWIW if there was a better way of visually marking this as dangerous I would.

**Reply by ibrahimqasim-abtrace** [reply]
> FWIW if there was a better way of visually marking this as dangerous I would.

I think that's all I was getting at. Since there's no other way to throw the warning in the users face, I can see your reasoning.

Thanks for all your help!
